// fast-fuzz/dist/src/fast-fuzz.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fuzzSkipProp = exports.fuzzSkipMethod = exports.fuzzSkipArg = exports.fuzzPropType = exports.fuzzProp = exports.fuzzMethod = exports.fuzzArgType = exports.fuzzArg = exports.fuzz = void 0;
require("reflect-metadata");
const fuzzCaller_1 = require("./fuzz/fuzzCaller");
Object.defineProperty(exports, "fuzz", { enumerable: true, get: function () { return fuzzCaller_1.fuzz; } });
const decorators_1 = require("./decorators");
Object.defineProperty(exports, "fuzzArg", { enumerable: true, get: function () { return decorators_1.arg; } });
Object.defineProperty(exports, "fuzzArgType", { enumerable: true, get: function () { return decorators_1.argType; } });
Object.defineProperty(exports, "fuzzMethod", { enumerable: true, get: function () { return decorators_1.method; } });
Object.defineProperty(exports, "fuzzProp", { enumerable: true, get: function () { return decorators_1.prop; } });
Object.defineProperty(exports, "fuzzPropType", { enumerable: true, get: function () { return decorators_1.propType; } });
Object.defineProperty(exports, "fuzzSkipArg", { enumerable: true, get: function () { return decorators_1.skipArg; } });
Object.defineProperty(exports, "fuzzSkipMethod", { enumerable: true, get: function () { return decorators_1.skipMethod; } });
Object.defineProperty(exports, "fuzzSkipProp", { enumerable: true, get: function () { return decorators_1.skipProp; } });
