diff --git a/node_modules/nodebb-plugin-composer-default/static/lib/composer.js b/node_modules/nodebb-plugin-composer-default/static/lib/composer.js
index 46334e7..19a480b 100644
--- a/node_modules/nodebb-plugin-composer-default/static/lib/composer.js
+++ b/node_modules/nodebb-plugin-composer-default/static/lib/composer.js
@@ -206,6 +206,7 @@ define('composer', [
 			tags: data.tags || [],
 			modified: !!((data.title && data.title.length) || (data.body && data.body.length)),
 			isMain: true,
+ 			isAnonymous: data.isAnonymous
 		};
 
 		({ pushData } = await hooks.fire('filter:composer.topic.push', {
@@ -431,6 +432,7 @@ define('composer', [
 		var isEditing = postData ? !!postData.pid : false;
 		var isGuestPost = postData ? parseInt(postData.uid, 10) === 0 : false;
 		const isScheduled = postData.timestamp > Date.now();
+		var isAnonymous = postData ? !!postData.isAnonymous : false;
 
 		// see
 		// https://github.com/NodeBB/NodeBB/issues/2994 and
@@ -454,6 +456,7 @@ define('composer', [
 			maximumTagLength: config.maximumTagLength,
 			isTopic: isTopic,
 			isEditing: isEditing,
+			isAnonymous : isAnonymous,
 			canSchedule: !!(isMain && privileges &&
 				((privileges['topics:schedule'] && !isEditing) || (isScheduled && privileges.view_scheduled))),
 			showHandleInput: config.allowGuestHandles &&
@@ -630,10 +633,12 @@ define('composer', [
 		var bodyEl = postContainer.find('textarea');
 		var thumbEl = postContainer.find('input#topic-thumb-url');
 		var onComposeRoute = postData.hasOwnProperty('template') && postData.template.compose === true;
+		var toggleBox = postContainer.find('.toggle-anonymous');
 		const submitBtn = postContainer.find('.composer-submit');
 
 		titleEl.val(titleEl.val().trim());
 		bodyEl.val(utils.rtrim(bodyEl.val()));
+		toggleBox.val(toggleBox[0].checked);
 		if (thumbEl.length) {
 			thumbEl.val(thumbEl.val().trim());
 		}
@@ -652,6 +657,7 @@ define('composer', [
 			titleLen: titleEl.val().length,
 			bodyEl: bodyEl,
 			bodyLen: bodyEl.val().length,
+			toggleBox: toggleBox,
 		};
 
 		await hooks.fire('filter:composer.check', payload);
@@ -696,6 +702,7 @@ define('composer', [
 				cid: categoryList.getSelectedCid(),
 				tags: tags.getTags(post_uuid),
 				timestamp: scheduler.getTimestamp(),
+				isAnonymous: toggleBox.val(),
 			};
 		} else if (action === 'posts.reply') {
 			route = `/topics/${postData.tid}`;
@@ -705,6 +712,7 @@ define('composer', [
 				handle: handleEl ? handleEl.val() : undefined,
 				content: bodyEl.val(),
 				toPid: postData.toPid,
+				isAnonymous: toggleBox.val(),
 			};
 		} else if (action === 'posts.edit') {
 			method = 'put';
@@ -718,6 +726,7 @@ define('composer', [
 				thumb: thumbEl.val() || '',
 				tags: tags.getTags(post_uuid),
 				timestamp: scheduler.getTimestamp(),
+				isAnonymous: toggleBox.val(),
 			};
 		}
 		var submitHookData = {
diff --git a/node_modules/nodebb-plugin-composer-default/static/templates/compose.tpl b/node_modules/nodebb-plugin-composer-default/static/templates/compose.tpl
index 8ce50b9..cbc8989 100644
--- a/node_modules/nodebb-plugin-composer-default/static/templates/compose.tpl
+++ b/node_modules/nodebb-plugin-composer-default/static/templates/compose.tpl
@@ -77,6 +77,8 @@
 					</form>
 				</ul>
 
+
+
 				<div class="btn-group pull-right action-bar">
 					<a href="{discardRoute}" class="btn btn-default composer-discard" data-action="discard" tabindex="-1"><i class="fa fa-times"></i> [[topic:composer.discard]]</a>
 
@@ -93,6 +95,7 @@
 				</div>
 				<textarea name="content" form="compose-form" class="write" tabindex="5" placeholder="[[modules:composer.textarea.placeholder]]">{body}</textarea>
 			</div>
+			<h1>CHICKEN MCNUGGETS </h1>
 			<div class="col-md-6 hidden-sm hidden-xs preview-container">
 				<div class="help-text">
 					<span class="toggle-preview">[[modules:composer.hide_preview]]</span>
diff --git a/node_modules/nodebb-plugin-composer-default/static/templates/composer.tpl b/node_modules/nodebb-plugin-composer-default/static/templates/composer.tpl
index cf9de24..a6427cf 100644
--- a/node_modules/nodebb-plugin-composer-default/static/templates/composer.tpl
+++ b/node_modules/nodebb-plugin-composer-default/static/templates/composer.tpl
@@ -13,6 +13,7 @@
 			<!-- ENDIF isTopic -->
 			<!-- IF !isTopicOrMain -->
 			<h4 class="title">[[topic:composer.replying_to, "{title}"]]</h4>
+			<label class="switch"><input type="checkbox"><span class="slider round"></span></label>
 			<!-- ENDIF !isTopicOrMain -->
 			<div class="display-scheduler pull-right{{{ if !canSchedule }}} hidden{{{ end }}}">
 				<i class="fa fa-clock-o"></i>
@@ -21,6 +22,7 @@
 				<button class="btn btn-sm btn-primary composer-submit" data-action="post" tabindex="-1"><i class="fa fa-chevron-right"></i></button>
 			</div>
 		</nav>
+		
 		<div class="row title-container">
 			{{{ if isTopic }}}
 			<div class="category-list-container hidden-sm hidden-xs">
@@ -56,6 +58,7 @@
 
 				<ul class="dropdown-menu">{{{ each submitOptions }}}<li><a href="#" data-action="{./action}">{./text}</a></li>{{{ end }}}</ul> 
 				<button class="btn btn-primary composer-submit" data-action="post" tabindex="6" data-text-variant=" [[topic:composer.schedule]]"><i class="fa fa-check"></i> [[topic:composer.submit]]</button>
+				<label class="switch"><input type="checkbox" class="toggle-anonymous"><span class="slider round">Post Anonymously</span></label>
 				<button type="button" class="btn btn-primary dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
 					<span class="caret"></span>
 					<span class="sr-only">[[topic:composer.additional-options]]</span>
@@ -120,6 +123,8 @@
 						[[tags:select_tags]]
 					</button>
 
+					
+
 					<ul class="dropdown-menu">
 						<!-- BEGIN tagWhitelist -->
 						<li data-tag="{@value}"><a href="#">{@value}</a></li>
