diff --git a/node_modules/nodebb-plugin-composer-default/static/less/composer.less b/node_modules/nodebb-plugin-composer-default/static/less/composer.less
index e9a8ca6..8df0e13 100644
--- a/node_modules/nodebb-plugin-composer-default/static/less/composer.less
+++ b/node_modules/nodebb-plugin-composer-default/static/less/composer.less
@@ -460,6 +460,66 @@
 			animation: draft-saved 3s ease;
 		}
 	}
+
+	.switch {
+		position: relative;
+		display: inline-block;
+		width: 60px;
+		height: 34px;
+	}
+		
+	.switch input { 
+		opacity: 0;
+		width: 0;
+		height: 0;
+	}
+		
+	.slider {
+		position: absolute;
+		cursor: pointer;
+		top: 0;
+		left: 0;
+		right: 0;
+		bottom: 0;
+		background-color: #ccc;
+		-webkit-transition: .4s;
+		transition: .4s;
+	}
+		
+	.slider:before {
+		position: absolute;
+		content: "";
+		height: 26px;
+		width: 26px;
+		left: 4px;
+		bottom: 4px;
+		background-color: white;
+		-webkit-transition: .4s;
+		transition: .4s;
+	}	
+		
+	input:checked + .slider {
+		background-color: #2196F3;
+	}
+		
+	input:focus + .slider {
+		box-shadow: 0 0 1px #2196F3;
+	}
+		
+	input:checked + .slider:before {
+		-webkit-transform: translateX(26px);
+		-ms-transform: translateX(26px);
+		transform: translateX(26px);
+	}
+		
+	/* Rounded sliders */
+	.slider.round {
+		border-radius: 34px;
+	}
+		
+	.slider.round:before {
+		border-radius: 50%;
+	}
 }
 
 .datetime-picker {
diff --git a/node_modules/nodebb-plugin-composer-default/static/lib/composer.js b/node_modules/nodebb-plugin-composer-default/static/lib/composer.js
index 46334e7..9592d58 100644
--- a/node_modules/nodebb-plugin-composer-default/static/lib/composer.js
+++ b/node_modules/nodebb-plugin-composer-default/static/lib/composer.js
@@ -197,6 +197,8 @@ define('composer', [
 	};
 
 	composer.newTopic = async (data) => {
+		console.log("========= Data =========")
+		console.log(data)
 		var pushData = {
 			action: 'topics.post',
 			cid: data.cid,
@@ -206,6 +208,7 @@ define('composer', [
 			tags: data.tags || [],
 			modified: !!((data.title && data.title.length) || (data.body && data.body.length)),
 			isMain: true,
+			isPrivate: data.isPrivate
 		};
 
 		({ pushData } = await hooks.fire('filter:composer.topic.push', {
@@ -426,11 +429,15 @@ define('composer', [
 	async function createNewComposer(post_uuid) {
 		var postData = composer.posts[post_uuid];
 
+		console.log("========== POST DATA ==========")
+		console.log(postData)
+
 		var isTopic = postData ? postData.hasOwnProperty('cid') : false;
 		var isMain = postData ? !!postData.isMain : false;
 		var isEditing = postData ? !!postData.pid : false;
 		var isGuestPost = postData ? parseInt(postData.uid, 10) === 0 : false;
 		const isScheduled = postData.timestamp > Date.now();
+		var isPrivate = postData ? !!postData.isPrivate : false;
 
 		// see
 		// https://github.com/NodeBB/NodeBB/issues/2994 and
@@ -454,6 +461,7 @@ define('composer', [
 			maximumTagLength: config.maximumTagLength,
 			isTopic: isTopic,
 			isEditing: isEditing,
+			isPrivate : isPrivate,
 			canSchedule: !!(isMain && privileges &&
 				((privileges['topics:schedule'] && !isEditing) || (isScheduled && privileges.view_scheduled))),
 			showHandleInput: config.allowGuestHandles &&
@@ -623,6 +631,7 @@ define('composer', [
 	}
 
 	async function post(post_uuid) {
+		console.log('post_uuid: ' + post_uuid)
 		var postData = composer.posts[post_uuid];
 		var postContainer = $('.composer[data-uuid="' + post_uuid + '"]');
 		var handleEl = postContainer.find('.handle');
@@ -631,9 +640,14 @@ define('composer', [
 		var thumbEl = postContainer.find('input#topic-thumb-url');
 		var onComposeRoute = postData.hasOwnProperty('template') && postData.template.compose === true;
 		const submitBtn = postContainer.find('.composer-submit');
+		var toggleEl = postContainer.find('.toggle');
+		console.log("HIIIIIIIIII!!!!!!!!")
+		console.log(postData)
+		// console.log('hi' + toggleEl[0].checked)
 
 		titleEl.val(titleEl.val().trim());
 		bodyEl.val(utils.rtrim(bodyEl.val()));
+		toggleEl.val(toggleEl[0].checked);
 		if (thumbEl.length) {
 			thumbEl.val(thumbEl.val().trim());
 		}
@@ -652,6 +666,7 @@ define('composer', [
 			titleLen: titleEl.val().length,
 			bodyEl: bodyEl,
 			bodyLen: bodyEl.val().length,
+			toggleEl: toggleEl,
 		};
 
 		await hooks.fire('filter:composer.check', payload);
@@ -696,6 +711,7 @@ define('composer', [
 				cid: categoryList.getSelectedCid(),
 				tags: tags.getTags(post_uuid),
 				timestamp: scheduler.getTimestamp(),
+				isPrivate: toggleEl.val(),
 			};
 		} else if (action === 'posts.reply') {
 			route = `/topics/${postData.tid}`;
diff --git a/node_modules/nodebb-plugin-composer-default/static/templates/compose.tpl b/node_modules/nodebb-plugin-composer-default/static/templates/compose.tpl
index 8ce50b9..4723d36 100644
--- a/node_modules/nodebb-plugin-composer-default/static/templates/compose.tpl
+++ b/node_modules/nodebb-plugin-composer-default/static/templates/compose.tpl
@@ -41,6 +41,10 @@
 			</div>
 		</div>
 
+		<div class="toggle-container private-toggle">
+			<!-- IMPORT partials/toggle.tpl -->
+		</div>
+
 		<div class="category-tag-row">
 			<div class="btn-toolbar formatting-bar">
 				<ul class="formatting-group">
diff --git a/node_modules/nodebb-plugin-composer-default/static/templates/composer.tpl b/node_modules/nodebb-plugin-composer-default/static/templates/composer.tpl
index cf9de24..35c375b 100644
--- a/node_modules/nodebb-plugin-composer-default/static/templates/composer.tpl
+++ b/node_modules/nodebb-plugin-composer-default/static/templates/composer.tpl
@@ -63,6 +63,10 @@
 			</div>
 		</div>
 
+		<div class="toggle-container private-toggle">
+			<!-- IMPORT partials/toggle.tpl -->
+		</div>
+
 		<div class="category-tag-row">
 			<div class="btn-toolbar formatting-bar">
 				<ul class="formatting-group">
