// fast-fuzz-shim/dist/src/decorators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.skipMethod = exports.method = exports.skipArg = exports.argType = exports.arg = exports.skipProp = exports.propType = exports.prop = void 0;
const prop = function (_type, _dimension = 0, _min, _max) {
    return (_target, _key) => {
        return;
    };
};
exports.prop = prop;
const propType = function (_typeName, _dimension = 0) {
    return (_target, _key) => {
        return;
    };
};
exports.propType = propType;
const skipProp = (_target, _key) => {
    return;
};
exports.skipProp = skipProp;
const arg = function (_type, _dimension = 0, _min, _max) {
    return (_target, _key, _index) => {
        return;
    };
};
exports.arg = arg;
const argType = function (_type, _dimension = 0) {
    return (_target, _key, _index) => {
        return;
    };
};
exports.argType = argType;
const skipArg = (_target, _key, _index) => {
    return;
};
exports.skipArg = skipArg;
const method = (_target, _key, descriptor) => {
    return descriptor;
};
exports.method = method;
const skipMethod = (_target, _key, descriptor) => {
    return descriptor;
};
exports.skipMethod = skipMethod;
