// fast-fuzz/dist/test/sut/enumDict.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Enum_Dict = exports.Direction_Mix_Auto = exports.Direction_Mix_Num = exports.Direction_Mix = exports.Direction_Num = void 0;
const fast_fuzz_1 = require("../../src/fast-fuzz");
const literals = {
    nil: 0,
    m1: -1,
    emptyStr: '',
    longStr: 'RkeRxOSF4BfSpgc5Dc7hGumMO',
    n1: 1,
    n2: 2,
    n3: 3,
    n4: 4,
    right: 'R'
};
var Direction_Num;
(function (Direction_Num) {
    Direction_Num[Direction_Num["Up"] = 0] = "Up";
    Direction_Num[Direction_Num["Down"] = 1] = "Down";
    Direction_Num[Direction_Num["Left"] = 3] = "Left";
    Direction_Num[Direction_Num["Right"] = 4] = "Right";
})(Direction_Num = exports.Direction_Num || (exports.Direction_Num = {}));
var Direction_Mix;
(function (Direction_Mix) {
    Direction_Mix[Direction_Mix["Up"] = 0] = "Up";
    Direction_Mix[Direction_Mix["Down"] = 1] = "Down";
    Direction_Mix[Direction_Mix["Left"] = 2] = "Left";
    Direction_Mix["Right"] = "R";
})(Direction_Mix = exports.Direction_Mix || (exports.Direction_Mix = {}));
var Direction_Mix_Num;
(function (Direction_Mix_Num) {
    Direction_Mix_Num[Direction_Mix_Num["Up"] = 0] = "Up";
    Direction_Mix_Num[Direction_Mix_Num["Down"] = 1] = "Down";
    Direction_Mix_Num[Direction_Mix_Num["Left"] = 3] = "Left";
    Direction_Mix_Num["Right"] = "R";
})(Direction_Mix_Num = exports.Direction_Mix_Num || (exports.Direction_Mix_Num = {}));
var Direction_Mix_Auto;
(function (Direction_Mix_Auto) {
    Direction_Mix_Auto[Direction_Mix_Auto["Up"] = 1] = "Up";
    Direction_Mix_Auto[Direction_Mix_Auto["Down"] = 2] = "Down";
    Direction_Mix_Auto[Direction_Mix_Auto["Left"] = 3] = "Left";
    Direction_Mix_Auto["Right"] = "R";
})(Direction_Mix_Auto = exports.Direction_Mix_Auto || (exports.Direction_Mix_Auto = {}));
class Enum_Dict {
    static enum_dict_num(arg) {
        if (arg === undefined) {
            return `Undefined: ${arg}`;
        }
        else if (arg === null) {
            return `Null: ${arg}`;
        }
        else if (Number.isNaN(arg)) {
            return `NaN: ${arg}`;
        }
        else if (arg === 0) {
            return `Zero: ${arg}`;
        }
        else if (arg === Number.MIN_SAFE_INTEGER) {
            return `Min int: ${arg}`;
        }
        else if (arg === Number.MAX_SAFE_INTEGER) {
            return `Max int: ${arg}`;
        }
        else if (arg === literals.m1 * Number.MIN_VALUE) {
            return `Negative min decimal: ${arg}`;
        }
        else if (arg === Number.MIN_VALUE) {
            return `Positive min decimal: ${arg}`;
        }
        else if (arg === literals.m1 * Number.MAX_VALUE) {
            return `Min float: ${arg}`;
        }
        else if (arg === Number.MAX_VALUE) {
            return `Max float: ${arg}`;
        }
        else if (arg === Number.NEGATIVE_INFINITY) {
            return `Negative infinity: ${arg}`;
        }
        else if (arg === Number.POSITIVE_INFINITY) {
            return `Positive infinity: ${arg}`;
        }
        else if (`${arg}` === literals.emptyStr) {
            return `Empty string: ${arg}`;
        }
        else if (`${arg}` === literals.longStr) {
            return `Long literal(25): ${arg}`;
        }
        else if (arg === literals.n1) {
            return `Direction.Down: ${arg}`;
        }
        else if (arg === literals.n3) {
            return `Direction.Left: ${arg}`;
        }
        else if (arg === literals.n4) {
            return `Direction.Right: ${arg}`;
        }
        else {
            return `Unkown: ${arg}`;
        }
    }
    static enum_dict_mix(arg) {
        if (arg === undefined) {
            return `Undefined: ${arg}`;
        }
        else if (arg === null) {
            return `Null: ${arg}`;
        }
        else if (Number.isNaN(arg)) {
            return `NaN: ${arg}`;
        }
        else if (arg === 0) {
            return `Zero: ${arg}`;
        }
        else if (arg === Number.MIN_SAFE_INTEGER) {
            return `Min int: ${arg}`;
        }
        else if (arg === Number.MAX_SAFE_INTEGER) {
            return `Max int: ${arg}`;
        }
        else if (arg === literals.m1 * Number.MIN_VALUE) {
            return `Negative min decimal: ${arg}`;
        }
        else if (arg === Number.MIN_VALUE) {
            return `Positive min decimal: ${arg}`;
        }
        else if (arg === literals.m1 * Number.MAX_VALUE) {
            return `Min float: ${arg}`;
        }
        else if (arg === Number.MAX_VALUE) {
            return `Max float: ${arg}`;
        }
        else if (arg === Number.NEGATIVE_INFINITY) {
            return `Negative infinity: ${arg}`;
        }
        else if (arg === Number.POSITIVE_INFINITY) {
            return `Positive infinity: ${arg}`;
        }
        else if (`${arg}` === literals.emptyStr) {
            return `Empty string: ${arg}`;
        }
        else if (`${arg}` === literals.longStr) {
            return `Long literal(25): ${arg}`;
        }
        else if (arg === literals.n1) {
            return `Direction.Down: ${arg}`;
        }
        else if (arg === literals.n2) {
            return `Direction.Left: ${arg}`;
        }
        else if (arg === literals.right) {
            return `Direction.Right: ${arg}`;
        }
        else {
            return `Unkown: ${arg}`;
        }
    }
    static enum_dict_mix_num(arg) {
        if (arg === undefined) {
            return `Undefined: ${arg}`;
        }
        else if (arg === null) {
            return `Null: ${arg}`;
        }
        else if (Number.isNaN(arg)) {
            return `NaN: ${arg}`;
        }
        else if (arg === 0) {
            return `Zero: ${arg}`;
        }
        else if (arg === Number.MIN_SAFE_INTEGER) {
            return `Min int: ${arg}`;
        }
        else if (arg === Number.MAX_SAFE_INTEGER) {
            return `Max int: ${arg}`;
        }
        else if (arg === literals.m1 * Number.MIN_VALUE) {
            return `Negative min decimal: ${arg}`;
        }
        else if (arg === Number.MIN_VALUE) {
            return `Positive min decimal: ${arg}`;
        }
        else if (arg === literals.m1 * Number.MAX_VALUE) {
            return `Min float: ${arg}`;
        }
        else if (arg === Number.MAX_VALUE) {
            return `Max float: ${arg}`;
        }
        else if (arg === Number.NEGATIVE_INFINITY) {
            return `Negative infinity: ${arg}`;
        }
        else if (arg === Number.POSITIVE_INFINITY) {
            return `Positive infinity: ${arg}`;
        }
        else if (`${arg}` === literals.emptyStr) {
            return `Empty string: ${arg}`;
        }
        else if (`${arg}` === literals.longStr) {
            return `Long literal(25): ${arg}`;
        }
        else if (arg === literals.n1) {
            return `Direction.Down: ${arg}`;
        }
        else if (arg === literals.n3) {
            return `Direction.Left: ${arg}`;
        }
        else if (arg === literals.right) {
            return `Direction.Right: ${arg}`;
        }
        else {
            return `Unkown: ${arg}`;
        }
    }
    static enum_dict_mix_auto(arg) {
        if (arg === undefined) {
            return `Undefined: ${arg}`;
        }
        else if (arg === null) {
            return `Null: ${arg}`;
        }
        else if (Number.isNaN(arg)) {
            return `NaN: ${arg}`;
        }
        else if (arg === literals.nil) {
            return `Zero: ${arg}`;
        }
        else if (arg === Number.MIN_SAFE_INTEGER) {
            return `Min int: ${arg}`;
        }
        else if (arg === Number.MAX_SAFE_INTEGER) {
            return `Max int: ${arg}`;
        }
        else if (arg === literals.m1 * Number.MIN_VALUE) {
            return `Negative min decimal: ${arg}`;
        }
        else if (arg === Number.MIN_VALUE) {
            return `Positive min decimal: ${arg}`;
        }
        else if (arg === literals.m1 * Number.MAX_VALUE) {
            return `Min float: ${arg}`;
        }
        else if (arg === Number.MAX_VALUE) {
            return `Max float: ${arg}`;
        }
        else if (arg === Number.NEGATIVE_INFINITY) {
            return `Negative infinity: ${arg}`;
        }
        else if (arg === Number.POSITIVE_INFINITY) {
            return `Positive infinity: ${arg}`;
        }
        else if (`${arg}` === literals.emptyStr) {
            return `Empty string: ${arg}`;
        }
        else if (`${arg}` === literals.longStr) {
            return `Long literal(25): ${arg}`;
        }
        else if (arg === literals.n1) {
            return `Direction.Up: ${arg}`;
        }
        else if (arg === literals.n2) {
            return `Direction.Down: ${arg}`;
        }
        else if (arg === literals.n3) {
            return `Direction.Left: ${arg}`;
        }
        else if (arg === literals.right) {
            return `Direction.Right: ${arg}`;
        }
        else {
            return `Unkown: ${arg}`;
        }
    }
}
__decorate([
    fast_fuzz_1.fuzzMethod,
    __param(0, (0, fast_fuzz_1.fuzzArgType)('Direction_Num')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Number]),
    __metadata("design:returntype", String)
], Enum_Dict, "enum_dict_num", null);
__decorate([
    fast_fuzz_1.fuzzMethod,
    __param(0, (0, fast_fuzz_1.fuzzArgType)('Direction_Mix')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", String)
], Enum_Dict, "enum_dict_mix", null);
__decorate([
    fast_fuzz_1.fuzzMethod,
    __param(0, (0, fast_fuzz_1.fuzzArgType)('Direction_Mix_Num')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", String)
], Enum_Dict, "enum_dict_mix_num", null);
__decorate([
    fast_fuzz_1.fuzzMethod,
    __param(0, (0, fast_fuzz_1.fuzzArgType)('Direction_Mix_Auto')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", String)
], Enum_Dict, "enum_dict_mix_auto", null);
exports.Enum_Dict = Enum_Dict;
