// fast-fuzz/dist/test/sut/skip.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Skip = void 0;
const fast_fuzz_1 = require("../../src/fast-fuzz");
const AGE_17 = 17;
const AGE_22 = 22;
class Skip {
    skip_instance_simple(_age) {
        switch (this.age) {
            case AGE_22:
                return false;
            case AGE_17:
                return false;
            default:
                break;
        }
        return true;
    }
    static skip_static_simple(_name, age, _code) {
        switch (age) {
            case AGE_22:
                return false;
            case AGE_17:
                return false;
            default:
                break;
        }
        return true;
    }
    static skip_static_skip(_name, age, _code) {
        switch (age) {
            case AGE_22:
                return false;
            case AGE_17:
                return false;
            default:
                break;
        }
        return true;
    }
    static skip_static_all(_name, age, _code) {
        switch (age) {
            case AGE_22:
                return false;
            case AGE_17:
                return false;
            default:
                break;
        }
        return true;
    }
    skip_instance_skip(_name, _age, _code) {
        this.private_prop = 1;
        switch (this.private_prop === 1 && this.age) {
            case AGE_22:
                return false;
            case AGE_17:
                return false;
            default:
                break;
        }
        return true;
    }
}
__decorate([
    (0, fast_fuzz_1.fuzzProp)("string"),
    __metadata("design:type", String)
], Skip.prototype, "prop_1", void 0);
__decorate([
    fast_fuzz_1.fuzzSkipProp,
    __metadata("design:type", Number)
], Skip.prototype, "age", void 0);
__decorate([
    (0, fast_fuzz_1.fuzzProp)("integer", 0, 5, 10),
    __metadata("design:type", Number)
], Skip.prototype, "code", void 0);
__decorate([
    fast_fuzz_1.fuzzMethod,
    __param(0, (0, fast_fuzz_1.fuzzArg)("integer", 0, 15, 25)),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Number]),
    __metadata("design:returntype", Boolean)
], Skip.prototype, "skip_instance_simple", null);
__decorate([
    fast_fuzz_1.fuzzSkipMethod,
    __param(0, fast_fuzz_1.fuzzSkipArg),
    __param(1, fast_fuzz_1.fuzzSkipArg),
    __param(2, fast_fuzz_1.fuzzSkipArg),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Number, Number]),
    __metadata("design:returntype", Boolean)
], Skip.prototype, "skip_instance_skip", null);
__decorate([
    fast_fuzz_1.fuzzMethod,
    __param(0, (0, fast_fuzz_1.fuzzArg)("string")),
    __param(1, fast_fuzz_1.fuzzSkipArg),
    __param(2, (0, fast_fuzz_1.fuzzArg)("integer", 0, 5, 10)),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Number, Number]),
    __metadata("design:returntype", Boolean)
], Skip, "skip_static_simple", null);
__decorate([
    fast_fuzz_1.fuzzSkipMethod,
    __param(0, (0, fast_fuzz_1.fuzzArg)("string")),
    __param(1, fast_fuzz_1.fuzzSkipArg),
    __param(2, (0, fast_fuzz_1.fuzzArg)("integer", 0, 5, 10)),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Number, Number]),
    __metadata("design:returntype", Boolean)
], Skip, "skip_static_skip", null);
__decorate([
    fast_fuzz_1.fuzzMethod,
    __param(0, fast_fuzz_1.fuzzSkipArg),
    __param(1, fast_fuzz_1.fuzzSkipArg),
    __param(2, fast_fuzz_1.fuzzSkipArg),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Number, Number]),
    __metadata("design:returntype", Boolean)
], Skip, "skip_static_all", null);
exports.Skip = Skip;
