// fast-fuzz/dist/src/intermock/build/test/ts/test-data/enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var GreatMusicians;
(function (GreatMusicians) {
    GreatMusicians["mozart"] = "Mozart";
    GreatMusicians["beethoven"] = "Beethoven";
})(GreatMusicians || (GreatMusicians = {}));
var GreatNovels;
(function (GreatNovels) {
    GreatNovels[GreatNovels["MOBY_DICK"] = 0] = "MOBY_DICK";
    GreatNovels[GreatNovels["GRAPES_OF_WRATH"] = 1] = "GRAPES_OF_WRATH";
    GreatNovels[GreatNovels["SLAUGHTERHOUSE_FIVE"] = 2] = "SLAUGHTERHOUSE_FIVE";
})(GreatNovels || (GreatNovels = {}));
var Awesomeness;
(function (Awesomeness) {
    Awesomeness[Awesomeness["COOL"] = 0] = "COOL";
    Awesomeness[Awesomeness["LAME"] = 1] = "LAME";
})(Awesomeness || (Awesomeness = {}));
var GreatNumbers;
(function (GreatNumbers) {
    GreatNumbers[GreatNumbers["e"] = 2.71] = "e";
    GreatNumbers[GreatNumbers["pi"] = 3.14] = "pi";
    GreatNumbers[GreatNumbers["golden"] = 1.61] = "golden";
})(GreatNumbers || (GreatNumbers = {}));
exports.expectedEnum = {
    Person: {
        name: 'Natasha Jacobs',
        status: 1,
        favoriteNumber: 3.14,
        favoriteMusicians: 'Beethoven',
        favoriteNovel: 1
    }
};
