// fast-fuzz/dist/src/utils/limits.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Limits = void 0;
class Limits {
    constructor(init) {
        const arg = {};
        for (const key in init) {
            const argValue = {};
            Object.assign(argValue, Limits.defaults[key]);
            if (init[key].min !== undefined) {
                argValue.min = init[key].min;
            }
            if (init[key].max !== undefined) {
                argValue.max = init[key].max;
            }
            if (key === 'date' ||
                key === 'string' ||
                key === 'array') {
                argValue.min = Math.max(argValue.min, 0);
            }
            if (key !== 'date' && argValue.max <= argValue.min) {
                argValue.max = argValue.min + Limits.defaults[key].max;
            }
            else if (key === 'date' && argValue.max <= argValue.min) {
                argValue.max = argValue.min + 8.64e+7;
            }
            arg[key] = argValue;
        }
        Object.assign(this, Limits.defaults, arg);
    }
}
exports.Limits = Limits;
Limits.defaults = {
    int: { min: -25, max: 25 },
    float: { min: -10, max: 10 },
    date: { min: Date.now() - 8.64e+7, max: Date.now() + 8.64e+7 },
    string: { min: 1, max: 4 },
    array: { min: 0, max: 3 }
};
